File: server/routes/mood.js
const express = require('express');
const router = express.Router();
const { verifyToken } = require('../middleware/auth');

// In-memory storage for demo (replace with Firestore in production)
const moodEntries = new Map();

const MOOD_TYPES = ['amazing', 'happy', 'calm', 'neutral', 'anxious', 'sad', 'angry', 'distressed'];

/**
 * POST /api/mood
 * Log a new mood entry.
 * Body: { mood: string, notes?: string }
 */
router.post('/', verifyToken, (req, res) => {
  try {
    const { mood, notes } = req.body;
    const userId = req.user.uid;

    if (!mood || !MOOD_TYPES.includes(mood)) {
      return res.status(400).json({
        error: `Invalid mood. Must be one of: ${MOOD_TYPES.join(', ')}`
      });
    }

    if (!moodEntries.has(userId)) {
      moodEntries.set(userId, []);
    }

    const entry = {
      id: Date.now().toString(),
      mood,
      notes: notes ? notes.trim() : '',
      timestamp: new Date().toISOString(),
      date: new Date().toISOString().split('T')[0]
    };

    moodEntries.get(userId).push(entry);

    res.json({
      success: true,
      data: entry,
      message: 'Mood logged successfully! 🌟'
    });
  } catch (error) {
    console.error('Mood logging error:', error);
    res.status(500).json({ error: 'Failed to log mood' });
  }
});

/**
 * GET /api/mood
 * Get mood history for the authenticated user.
 * Query: ?days=7 (default 30)
 */
router.get('/', verifyToken, (req, res) => {
  try {
    const userId = req.user.uid;
    const days = parseInt(req.query.days) || 30;
    const entries = moodEntries.get(userId) || [];

    // Filter by date range
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - days);

    const filtered = entries.filter(e => new Date(e.timestamp) >= cutoffDate);

    // Calculate mood distribution
    const distribution = {};
    MOOD_TYPES.forEach(m => { distribution[m] = 0; });
    filtered.forEach(e => {
      distribution[e.mood] = (distribution[e.mood] || 0) + 1;
    });

    res.json({
      success: true,
      data: {
        entries: filtered,
        distribution,
        totalEntries: filtered.length,
        dateRange: { from: cutoffDate.toISOString(), to: new Date().toISOString() }
      }
    });
  } catch (error) {
    console.error('Mood history error:', error);
    res.status(500).json({ error: 'Failed to retrieve mood history' });
  }
});

/**
 * GET /api/mood/today
 * Check if the user has logged a mood today.
 */
router.get('/today', verifyToken, (req, res) => {
  try {
    const userId = req.user.uid;
    const entries = moodEntries.get(userId) || [];
    const today = new Date().toISOString().split('T')[0];

    const todayEntry = entries.find(e => e.date === today);

    res.json({
      success: true,
      data: {
        hasLoggedToday: !!todayEntry,
        entry: todayEntry || null
      }
    });
  } catch (error) {
    console.error('Today mood check error:', error);
    res.status(500).json({ error: 'Failed to check today\'s mood' });
  }
});

module.exports = router;

File: backend/routes/mood.py
"""
Mood tracking routes.
Uses in-memory storage with Supabase persistence when available.
"""

from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends, HTTPException, Query

from middleware.auth import get_current_user, CurrentUser
from models.schemas import MoodRequest, VALID_MOODS
from config.supabase import get_supabase

router = APIRouter()

# In-memory fallback
_mood_entries: dict[str, list[dict]] = {}


@router.post("/")
async def log_mood(
    body: MoodRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """Log a new mood entry."""
    if body.mood not in VALID_MOODS:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid mood. Must be one of: {', '.join(VALID_MOODS)}",
        )

    now = datetime.now(timezone.utc)
    entry = {
        "id": str(int(now.timestamp() * 1000)),
        "mood": body.mood,
        "notes": (body.notes or "").strip(),
        "timestamp": now.isoformat(),
        "date": now.strftime("%Y-%m-%d"),
    }

    # Persist to Supabase
    sb = get_supabase()
    if sb:
        try:
            result = sb.table("mood_entries").insert({
                "user_id": user.uid,
                "mood": body.mood,
                "notes": entry["notes"],
            }).execute()
            if result.data:
                entry["id"] = str(result.data[0]["id"])
        except Exception:
            pass

    # In-memory fallback
    if user.uid not in _mood_entries:
        _mood_entries[user.uid] = []
    _mood_entries[user.uid].append(entry)

    return {
        "success": True,
        "data": entry,
        "message": "Mood logged successfully!",
    }


@router.get("/")
async def get_mood_history(
    user: CurrentUser = Depends(get_current_user),
    days: int = Query(default=30, ge=1, le=365),
):
    """Get mood history for the authenticated user."""
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)

    sb = get_supabase()
    entries = []

    if sb:
        try:
            result = (
                sb.table("mood_entries")
                .select("*")
                .eq("user_id", user.uid)
                .gte("created_at", cutoff.isoformat())
                .order("created_at", desc=False)
                .execute()
            )
            if result.data:
                entries = [
                    {
                        "id": str(row["id"]),
                        "mood": row["mood"],
                        "notes": row.get("notes", ""),
                        "timestamp": row["created_at"],
                        "date": row["created_at"][:10],
                    }
                    for row in result.data
                ]
        except Exception:
            pass

    # Fallback to in-memory
    if not entries:
        all_entries = _mood_entries.get(user.uid, [])
        entries = [
            e for e in all_entries
            if datetime.fromisoformat(e["timestamp"]) >= cutoff
        ]

    # Calculate distribution
    distribution = {m: 0 for m in VALID_MOODS}
    for e in entries:
        distribution[e["mood"]] = distribution.get(e["mood"], 0) + 1

    return {
        "success": True,
        "data": {
            "entries": entries,
            "distribution": distribution,
            "totalEntries": len(entries),
            "dateRange": {
                "from": cutoff.isoformat(),
                "to": datetime.now(timezone.utc).isoformat(),
            },
        },
    }


@router.get("/today")
async def get_today_status(user: CurrentUser = Depends(get_current_user)):
    """Check if the user has logged a mood today."""
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")

    sb = get_supabase()
    if sb:
        try:
            result = (
                sb.table("mood_entries")
                .select("*")
                .eq("user_id", user.uid)
                .gte("created_at", f"{today}T00:00:00")
                .limit(1)
                .execute()
            )
            if result.data:
                row = result.data[0]
                return {
                    "success": True,
                    "data": {
                        "hasLoggedToday": True,
                        "entry": {
                            "id": str(row["id"]),
                            "mood": row["mood"],
                            "notes": row.get("notes", ""),
                            "timestamp": row["created_at"],
                            "date": row["created_at"][:10],
                        },
                    },
                }
        except Exception:
            pass

    # In-memory fallback
    all_entries = _mood_entries.get(user.uid, [])
    today_entry = next((e for e in reversed(all_entries) if e["date"] == today), None)

    return {
        "success": True,
        "data": {
            "hasLoggedToday": today_entry is not None,
            "entry": today_entry,
        },
    }
